env.c

```diff
@@ -19,11 +19,19 @@ static struct treenode *pools;
 
 static void addpool(struct pool *);
 
+static void delpool(void *);
+static void delrule(void *);
+
 void
 envinit(void)
 {
-	/* TODO: delete old root environment and pools (in case we rebuilt
-	 * build.ninja). for now, we leak memory. */
+	/* free old rootenv in case rebuilt the manifest */
+	if (rootenv) {
+		deltree(rootenv->bindings, free);
+		deltree(rootenv->rules, delrule);
+		deltree(pools, delpool);
+		free(rootenv);
+	}
```

Comment:
I think we still leak memory if `subninja` was used.

Fixing this would probably involve keeping a list of all environments, and traversing the list here. What do you think?

Reply:
I haven't read the whole codebase yet, so I don't know which environments I need to remove. Also no idea what `subninja` is. I'll write a TODO for the extra leaked memory, but keep the code as is.

Reply:
The root environment that contains definitions in `build.ninja`. If `build.ninja` contains any [`subninja`](https://ninja-build.org/manual.html#ref_scope) lines, new environments will be created with the root environment as a parent, and so on.

Since we only have pointers back to the parent, there is currently no way to traverse the tree of environments. I can fix this in a follow-up commit. A TODO is fine for now.